File: c++/theory/rectangle.cpp
#include "shape.cpp"

class rectangle : public shape {
private:
    //Field (Attributi) Stato interno dell'oggetto
    int length; 
    int width;

public:
    //Costruttori:
    //Inizializzazione di oggetti della classe
    // = stato iniziale di questi oggetti
    rectangle() {
        length = 0;
        width = 0;
    }

    rectangle(int length, int width) {
        this->length = length;
        this->width = width;
    }

    //Methods (metodi) -> Servizi
    int perimeter() {
        return 2 * length + 2 * width;
    }

    double area() {
        return length * width;
    }
    //...
};

int main() {

    //Call super Method through scope resolution
    rectangle rect;
    rect.shape::area();

    return 0;
}

File: java/net/davoleo/unipr/Rectangle.java
package net.davoleo.unipr;

public class Rectangle {
    protected int width;
    protected int height;
    
    public Rectangle() {
        width = 0;
        height = 0;
    }
    
    public Rectangle(int width, int height) {
        this.width = width;
        this.height = height;
    }
    
    public int perimeter() {
        return 2 * width + 2 * height;
    }
    
    public int area() {
        return height * width;
    }

    public static class Square extends Rectangle {
        public Square() {}

        public Square(int size) {
            super(size, size);
        }

        public double getInscribedCircleLength() {
            return height * Math.PI;
        }

        @Override
        public int perimeter() {
            return 4 * height;
        }
    }
}